File: lib/lambda/rules/aws_cis/aws_cis_3_10.ts
import { cloudTrailRecord, uguisuRule, detection } from "../../models";

// https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html#cis-3.10-remediation

export class rule extends uguisuRule {
  constructor() {
    super({
      id: "aws_cis_3.10",
      title: "Security group changes",
      description:
        "AWS CIS benchmark 3.10 recommend to ensure a log metric filter and alarm exist for security group changes",
      severity: "medium",
    });
  }

  detect(record: cloudTrailRecord): detection | null {
    const targetEvents: { [key: string]: boolean } = {
      AuthorizeSecurityGroupIngress: true,
      AuthorizeSecurityGroupEgress: true,
      RevokeSecurityGroupIngress: true,
      RevokeSecurityGroupEgress: true,
      CreateSecurityGroup: true,
      DeleteSecurityGroup: true,
    };

    if (targetEvents[record.eventName] === true) {
      return {
        rule: this,
        event: record,
      };
    }

    return null;
  }
}

File: pkg/rules/aws_cis_3_10.go
package rules

import (
	"github.com/m-mizutani/uguisu/pkg/models"
)

// https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html#cis-3.10-remediation
type awsCIS3_10 struct {
	targetEvents map[string]bool
}

func newAwsCIS3_10() models.Rule {
	return &awsCIS3_10{
		targetEvents: map[string]bool{
			"AuthorizeSecurityGroupIngress": true,
			"AuthorizeSecurityGroupEgress":  true,
			"RevokeSecurityGroupIngress":    true,
			"RevokeSecurityGroupEgress":     true,
			"CreateSecurityGroup":           true,
			"DeleteSecurityGroup":           true,
		},
	}
}

func (x *awsCIS3_10) ID() string                { return "aws_cis_3.10" }
func (x *awsCIS3_10) Title() string             { return "Security group changes" }
func (x *awsCIS3_10) Severity() models.Severity { return models.SeverityMedium }
func (x *awsCIS3_10) Description() string {
	return "AWS CIS benchmark 3.10 recommend to ensure a log metric filter and alarm exist for security group changes"
}
func (x *awsCIS3_10) Match(record *models.CloudTrailRecord) bool {
	return x.targetEvents[record.EventName] == true
}
